control-plane/agents/src/bin/core/pool/registry.rs

```diff
@@ -3,10 +3,75 @@ use crate::controller::{
     wrapper::{GetterOps, *},
 };
 use agents::errors::{SvcError, SvcError::PoolNotFound};
-use stor_port::types::v0::transport::{CtrlPoolState, NodeId, Pool, PoolId, Replica, ReplicaId};
+use std::collections::HashMap;
+use stor_port::types::v0::transport::{
+    CtrlPoolState, Labels, NodeId, Pool, PoolId, Replica, ReplicaId,
+};
 
 /// Pool helpers
 impl Registry {
+    pub(crate) async fn get_label_opts_pools(
+        &self,
+        labels: Option<Labels>,
+    ) -> Result<Vec<Pool>, SvcError> {
+        let mut pools = vec![];
+        let pools_from_state = self.ctrl_pool_states().await.into_iter().map(|state| {
+            let spec = self.specs().pool(&state.id).ok();
+            Pool::from_state(state, spec)
+        });
+
+        pools.extend(pools_from_state);
+
+        let pools_from_spec = self
+            .specs()
+            .pools()
+            .into_iter()
+            .filter(|p| !pools.iter().any(|i| i.id() == &p.id))
+            .map(Pool::from_spec)
+            .collect::<Vec<_>>();
+
+        pools.extend(pools_from_spec);
+
+        let pools_after = match labels {
+            Some(label) => {
+                let mut filtered_pool: Vec<Pool> = vec![];
+                for pool in pools.iter() {
+                    match pool.spec() {
+                        Some(spec) => match spec.labels {
+                            Some(pool_labels) => {
+                                let pool_label_match =
+                                    Self::labels_matched(pool_labels, label.clone())?;
+                                if pool_label_match {
+                                    filtered_pool.push(pool.clone());
+                                }
+                            }
+                            None => continue,
+                        },
+                        None => continue,
+                    }
+                }
+                return Ok(filtered_pool);
+            }
+            None => pools,
+        };
+        Ok(pools_after)
+    }
+
+    pub(crate) fn labels_matched(
+        pool_labels: HashMap<String, String>,
+        labels: Labels,
+    ) -> Result<bool, SvcError> {
+        for label in labels.split(',') {
```

Comment:
I'm confused, doesn't the hashmap already have key and value? Why do we need further parsing?

Reply:
no, the query params comes as a string lie `A=B,C=D` and here it is parses. The labels of the pool are a hashmap

Reply:
Looks like that's an openapi limitation due to using query rather than body. 
Let's change it from string to hashmap on the grpc proto.